tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -421,3 +433,129 @@ def prepare_ion_recomb_coefficients_nlte_ion(
             total_coll_ion_coefficients,
             total_coll_recomb_coefficients,
         )
+
+    @staticmethod
+    def jacobian_matrix(
+        populations,
+        rate_matrix,
+        rate_matrix_index,
+        total_rad_recomb_coefficients,
+        total_coll_ion_coefficients,
+        total_coll_recomb_coefficients,
+    ):
+        """Creates the jacobian matrix used for NLTE ionization solver
+
+        Parameters
+        ----------
+        populations : numpy.array
+            Ion populations, electron density
+        rate_matrix : DataFrame
+            Rate matrix used for NLTE solver.
+        rate_matrix_index : MultiIndex
+            (atomic_number, ion_number, treatment type)
+            If ion is treated in LTE or nebular ionization, 3rd index is "lte_ion",
+            if treated in NLTE ionization, 3rd index is "nlte_ion".
+        total_rad_recomb_coefficients : DataFrame
+            Radiative recombination coefficients grouped by atomic number and ion number.
+        total_coll_ion_coefficients : DataFrame
+            Collisional ionization coefficients(should get multiplied by electron density).
+        total_coll_recomb_coefficients : DataFrame
+            Collisional recombination coefficients(should get multiplied by electron density).
+
+        Returns
+        -------
+        numpy.array
+            Jacobian matrix used for NLTE ionization solver
+        """
+        # TODO: for future use, can be vectorized.
+        atomic_numbers = (
+            rate_matrix_index.get_level_values("atomic_number")
+            .unique()
+            .drop("n_e")
+        )
```

Comment:
I think this is used multiple times. Maybe calculate it once at the beginning of the calculate method and set it as an attribute. In principle, you can also make the atomic numbers an input of the Solver class since they do not change.